Grid class restructuring

mineserver228 opened the pull request.

mineserver228 committed:
Grid class optimization

- Deleted functions
  - loadEntropy - it was really laggy

- Added functions
  - loadPosEntropy - loads entropy and saves it at pos in grid
  - loadEntropyAroundPos - loads entropy around pos

- Changed functions
  - setRandom - now it loads entropy and not saving image
  - collapse - optimized and debugged, before it was just deleting image
    instead of making backups

kolayne commented on wfcGen.py:
```diff
@@ -1,7 +1,10 @@
 import random
+import time
```
Why?

mineserver228 replied:
needed for debug, to see wich part takes most of the time and needs optimization, forgot to delete

kolayne commented on wfcGen.py:
```diff
@@ -164,56 +167,63 @@ def __init__(self, sx, sy, cells, tileSize=3):
+        self.loadEntropyAroundPos([rand[1], rand[0]])
+
+    def loadPosEntropy(self, pos):
+        for entropy, value in self.entropySave.items():
```
I suppose there is a better way to name `value`. What does this variable really hold?

mineserver228 replied:
coordinats of cells in grid wich have this entropy

kolayne commented on wfcGen.py:
```diff
@@ -164,56 +167,63 @@ def __init__(self, sx, sy, cells, tileSize=3):
+        self.entropySave[entropy] = self.entropySave.get(entropy, []) + [[pos[0], pos[1]]]
+    
+    def loadEntropyAroundPos(self, pos):
+        for x in [-1, 0, 1]:
+                for y in [-1, 0, 1]:
+                    if abs(x) == abs(y):
+                        continue
+                    if (pos[1]+y)<len(self.grid) and (pos[0]+x)<len(self.grid):
+                        self.loadPosEntropy([pos[0]+x, pos[1]+y])
```
Indentation issue

mineserver228 replied:
sorry, I think it is caused by copy pasting from another place in code)

kolayne reviewed (changes requested):
Nice, looks like there are some positive changes!

1. Please, fix a few minor issues in the code.
2. Please, consider first getting #2 merged, as you checked out this branch from the tip of the #2's branch, so the current PR includes all commits from #2. I would recommend that you resolve problems from #2 first

mineserver228 committed:
deleted unnecesary code and fixed indentation

- Changed functions
  - loadEntropyAroundPos - fixed inndentation issue
  - loadPosEntropy - renamed variables
    value -> cellPoses

- deleted unnecesary import

mineserver228 committed:
new function imagePixelsToArray and bug fixes

- Added functions
  - imagePixelsToArray - makes array of id's out of image

- Fixed bugs
  - in loadEntropyAroundPos - sometimes it was giving incorrect x and y
    to loadPosEntropy function

mineserver228 committed:
Merge branch 'main' into GridClassRestructuring

kolayne commented on wfcGen.py:
```diff
@@ -222,22 +229,22 @@ def collapse(self):
         self.GIMAGE.save(saveFileName)
-
+    
+    def imagePixelsToArray(self, colorToPixelDict):
+        return [[colorToPixelDict[self.GIMAGE.getpixel((x, y))] if self.GIMAGE.getpixel((x, y)) in colorToPixelDict.keys() else -1 for x in range(self.sx*self.tileSize)] for y in range(self.sy*self.tileSize)]
```
I'm pretty sure there is a more readable way to write this but ok. As long as this is a very short function targeting a very narrow purpose with the good function name and parameter name, I'm gonna accept it.

mineserver228 committed:
Grid optimization, bug fix and code restructuring

- Changed functions
  - loadPosEntropy - optimized, now it is working 2.3 times faster

- New variables
  - self.invertedEntropy - saves keys as data, and data as keys in
    self.entropySave

- New features
  - Timer - tells how much time generation takes, doesn't count image
    loading and cells creation

- Code changes
  - cells is now not dict, it is array now

kolayne commented on wfcGen.py:
```diff
@@ -164,85 +166,97 @@ def __init__(self, sx, sy, cells, tileSize=3):
-                if not self.grid[y][x].collapsed:
-                    entropy = self.grid[y][x].getEntropy(self.cells)
-                    self.entropySave[entropy] = self.entropySave.get(entropy, []) + [[x, y]]
+        rand = [random.randint(0, self.sy-1), random.randint(0, self.sx-1)]
```
This variable does not make much sense. As far as I can see, you are only using `rand[0]` and `rand[1]` separately. So, why not just have two variables (with telling names!)?

kolayne commented on wfcGen.py:
```diff
@@ -164,85 +166,97 @@ def __init__(self, sx, sy, cells, tileSize=3):
+        self.loadPosEntropy([rand[1], rand[0]])
+        self.loadEntropyAroundPos([rand[1], rand[0]])
+
+    def loadPosEntropy(self, pos):
+        if (pos[0], pos[1]) in list(self.invertedEntropy.keys()) and pos in self.entropySave[self.invertedEntropy[(pos[0], pos[1])]]:
+            self.entropySave[self.invertedEntropy[(pos[0], pos[1])]].remove(pos)
+            if self.entropySave[self.invertedEntropy[(pos[0], pos[1])]] == []:
+                self.entropySave.pop(self.invertedEntropy[(pos[0], pos[1])], None)
+        self.grid[pos[1]][pos[0]].loadEntropy(self.grid)
+        entropy = self.grid[pos[1]][pos[0]].getEntropy(self.cells)
+        if entropy != -1:
+            self.entropySave[entropy] = self.entropySave.get(entropy, []) + [pos]
+            self.invertedEntropy[(pos[0], pos[1])] = entropy
+        else:
+            self.invertedEntropy.pop((pos[0], pos[1]), None)
```
Here, as well, in most places you use `pos[0]` and `pos[1]` as two separate values. Does it make sense to unpack?
```python
x, y = pos
```

kolayne replied:
Btw, here you may have the temptation to do something like this:
```python
def loadPosEntropy(self, pos):
    x, y = pos

    # ...

    # When you need them separately:
    do_something_with(x, y)

    # When you need them together
    do_together(pos)
```

However, I would highly recommend to avoid that, as the code makes it appear as if `x, y` and `pos` are two separate entities, while in fact they are the same thing in different forms. To make the code more clear, it is better to only use one of them, as follows:
```python
def loadPosEntropy(self, pos):
    x, y = pos  # Once unpacked, `pos` is never used again

    # ...

    # When you need them separately:
    do_something_with(x, y)

    # When you need them together
    do_together((x, y))
```

kolayne commented on wfcGen.py:
```diff
@@ -164,85 +166,97 @@ def __init__(self, sx, sy, cells, tileSize=3):
+        else:
+            self.invertedEntropy.pop((pos[0], pos[1]), None)
+    
+    def loadEntropyAroundPos(self, pos):
```
Same comment about the use of `pos` as for `loadPosEntropy`

kolayne commented on wfcGen.py:
```diff
@@ -164,85 +166,97 @@ def __init__(self, sx, sy, cells, tileSize=3):
-        y = self.entropySave[sortedEntropy[0]][0][1]
-        x = self.entropySave[sortedEntropy[0]][0][0]
+        entropyPos = random.choice(self.entropySave[sortedEntropy[0]])
+        y = entropyPos[1]
+        x = entropyPos[0]
```
Is it `x, y = entropyPos`?

Let me remind you, unpacking a tuple instead of indexing its members gives the advantage of a type check that ensures the number of members:

```python
def unpack1(coords3):
    x, y, z = coords3

def unpack2(coords3):
    x = coords3[0]
    y = coords3[1]
    z = coords3[2]

if __name__ == '__main__':
    # Successful either way:
    pos = (1, 2, 3)
    unpack1(pos)
    unpack2(pos)

    # Exception raised either way:
    pos = (1, 2)
    unpack1(pos)  # Raises `TypeError`
    unpack2(pos)  # Raises `IndexError`

    # Only one exception!
    pos = (1, 2, 3, 4)
    unpack1(pos)  # Raises `TypeError`, so you know something is wrong
    unpack2(pos)  # Silently ignores the problem, you don't get to know anything was wrong
```

In cases when you really need to extract only a few members of a tuple, there is an option to do this with unpacking too:
```python
# Simple case: want to extract just `1` and `3`
redundant = (1, 2, 3, 4)
x, _, y, _ = redundant  # Yay!

# Nested case: want to extract `2` and `4`
redundant = (1, (2, 3), 4, (5,))
_, (x, _), y, (_,) = redundant  # Yay!
```

kolayne commented on wfcGen.py:
```diff
@@ -164,85 +166,97 @@ def __init__(self, sx, sy, cells, tileSize=3):
-            for j in range(self.reloads + random.randint(0, 4)):
+            for i in range(self.reloads + random.randint(0, 3)):
                 if len(self.backUps) > 0:
                     lb = self.backUps[len(self.backUps)-1]
```
Should this be unpacked too?

mineserver228 replied:
what unpacked?
I don't see anything there to unpack

kolayne replied:
I mean `lb`. Have a look at how you use `lb` later. It seems like you mostly need the two components of it, rather than the whole tuple

kolayne reviewed (changes requested):
Firstly, there were some changes I requested last time that you did not fix yet. Secondly, here are some more

mineserver228 committed:
Bug fix and code look changes

- Changed unpack method/Unpacked
  - loadPosEntropy
    pos -> posx, posy
  - loadEntropyAroundPos
    pos -> posx, posy
  - collapse
    entropyPos -> x, y
  - setRandom
    rand -> randomXPos, randomYPos

- fixed small bug
